File: release/push-tag.rb
#!/usr/bin/env ruby

require_relative 'common'

module Itch
  def Itch.ci_push_tag
    force = false

    pkg_path = File.expand_path(File.join(
      File.dirname(__FILE__), '..', 'package.json'))

    pkg = JSON.parse(File.read(pkg_path))
    pkg_version = pkg['version']

    force = false
    args = []
    ARGV.each do |arg|
      case arg
      when "--force"
        say "(Running in forced mode)"
        force = true
      when /^--/
        raise "Unknown option #{arg}"
      else
        args << arg
      end
    end

    version_input = args[0] or prompt "Package version is: #{pkg['version']}, type yours"
    unless /^v\d+.\d+.\d+(-canary)?$/ =~ version_input
      raise "Version must be of the form vX.Y.Z(-canary)?"
    end

    next_version = version_input.gsub(/^v/, '')

    if pkg_version != next_version
      force or yesno "Bump package.json? [#{pkg_version} => #{next_version}]"
      pkg['version'] = next_version
      File.write(pkg_path, JSON.pretty_generate(pkg))
      say "Bumped package.json"
      ✓ sh %Q{git add package.json}
      ✓ sh %Q{git commit -m ':arrow_up: #{next_version}'}
    end

    tag = "v#{next_version}"
    add_cmd = %Q{git tag -s -a #{tag} -m #{tag}}

    if system add_cmd
      say "Tag added..."
    else
      force or yesno "Tag already exists locally. Replace?"
      ✓ sh %Q{git tag -d #{tag}}
      ✓ sh add_cmd
    end

    push_cmd = %Q{git push origin #{tag}}
    if system push_cmd
      say "Tag pushed..."
    else
      force or yesno "Tag already exists on remote. Force-push?"
      ✓ sh %Q{#{push_cmd} --force}
    end

    ✓ sh %Q{git push}
  end
end

Itch.ci_push_tag

File: release/push-tag.js
//@ts-check
"use strict";

const { $, prompt, confirm } = require("@itchio/bob");
const { readFileSync, writeFileSync } = require("fs");
const { resolve } = require("path");

async function pushTag() {
  const pkgPath = resolve(__dirname, "..", "package.json");
  const pkg = JSON.parse(readFileSync(pkgPath, { encoding: "utf-8" }));
  const pkgVersion = pkg.version;

  let force = false;
  const args = [];
  for (const arg of process.argv.slice(2)) {
    switch (arg) {
      case "--force":
        console.log("(Running in forced mode)");
        force = true;
        break;
      default:
        if (/^--/.test(arg)) {
          throw new Error(`Unknown option ${arg}`);
        }
        args.push(arg);
    }
  }

  const versionInput =
    args[0] || (await prompt(`Package version is: ${pkg.version}, type yours in vX.Y.Z(-canary?) format`));
  if (!/^v\d+.\d+.\d+(-canary)?$/.test(versionInput)) {
    throw new Error(
      `Version must be of the form /vX.Y.Z(-canary)?/ (was '${versionInput}')`
    );
  }

  const nextVersion = versionInput.replace(/^v/, "");

  if (pkgVersion !== nextVersion) {
    if (!force) {
      await confirm(`Bump package.json? [${pkgVersion} => ${nextVersion}]`);
    }
    pkg.version = nextVersion;
    writeFileSync(pkgPath, JSON.stringify(pkg, null, 2), { encoding: "utf-8" });
    console.log("Bumped package.json");
    console.log("Letting npm bump package-lock.json...");
    $(`npm i`);
    $(`git add package.json package-lock.json`);
    // note: `--no-verify` works around a lint-staged@10.2.13 path bug
    // on msys2 on Windows.
    $(`git commit -m ':arrow_up: ${nextVersion}' --no-verify`);
  }

  const tag = `v${nextVersion}`;
  const isCanary = /-canary$/.test(nextVersion);
  const addCmd = `git tag ${isCanary ? "" : "-s"} -a ${tag} -m ${tag}`;

  try {
    $(addCmd);
    console.log("Tag added...");
  } catch (e) {
    if (!force) {
      await confirm("Tag already exists locally. Replace?");
    }
    $(`git tag -d ${tag}`);
    $(addCmd);
  }

  const pushCmd = `git push origin ${tag}`;
  try {
    $(pushCmd);
    console.log("Tag pushed...");
  } catch (e) {
    if (!force) {
      confirm("Tag already exists on remote. Force-push?");
    }
    $(`${pushCmd} --force`);
  }

  $("git push");
}

pushTag();
